### PR TITLE
Av 1058 continuity

### PR DESCRIPTION
**Pull request by zyunlam**

### CODE REVIEW

File: MIDAS/src/hardware/Continuity.cpp [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
-// global static instance of the sensor
+#include <hal.h>
 
+ADS7138_1Q ads7138_1q;
 
 ErrorCode ContinuitySensor::init() {
     // do whatever steps to initialize the sensor
     // if it errors, return the relevant error code
+    //pinMode(input_pin, INPUT);
+    //pinMode(output_pin, OUTPUT);
```
Comment (aidancostello) [inline MIDAS/src/hardware/Continuity.cpp:12]: create a placeholder init function for the adc and call it here

File: MIDAS/src/hardware/drivers/ads7138-q1.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#pragma once
```
Comment (aidancostello) [inline MIDAS/src/hardware/drivers/ads7138-q1.h:1]: move this whole file to the lib folder. Follow the file structure in lib. you can rebase off of mvp-sensor-integration to get this folder back, I think it got deleted somehow.

File: MIDAS/src/hardware/drivers/ads7138-q1.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+#include <stdint.h>
+// Driver todo
+// https://www.ti.com/product/ADS7138-Q1
+
```
Comment (aidancostello) [inline MIDAS/src/hardware/drivers/ads7138-q1.h:6]: add an init function that returns a success/fail

File: MIDAS/src/main.cpp [hunk 4]
```diff
@@ -5,6 +5,15 @@
  */
 RocketSystems systems;
 
+#ifdef HILSIM
+#include "hilsim/hilsim.pb.h"
```
Comment (aidancostello) [inline MIDAS/src/main.cpp:9]: remove this hilsim stuff for now it's not necessary for this pr

File: MIDAS/src/sensor_data.h [hunk 5]
```diff
@@ -63,8 +63,19 @@ struct Barometer {
...
 };
 
+// Detects if a charge or 
 struct Continuity {
-    bool is_continuous = false;
+    Continuity() = default;
+    bool pins[6];
+    // fancy operator overloading, you CS 128 kids won't get it until like week 6, and even then I don't think y'all
+    // mess with overloading []
+    bool& operator[](const int i) {
```
Comment (aidancostello) [inline MIDAS/src/sensor_data.h:72]: I don't think we need this overloading, we can just access the pins field itself. It seems a little overcomplicated

File: MIDAS/src/sensor_data.h [hunk 6]
```diff
@@ -63,8 +63,19 @@ struct Barometer {
     float pressure = 0;
 };
 
+// Detects if a charge or 
 struct Continuity {
-    bool is_continuous = false;
+    Continuity() = default;
```
Comment (aidancostello) [inline MIDAS/src/sensor_data.h:68]: remove the default constructor, just initialize it with the {} in the update function

File: MIDAS/src/sensors.h [hunk 7]
```diff
@@ -26,6 +26,10 @@ struct BarometerSensor {
 struct ContinuitySensor {
     ErrorCode init();
     Continuity read();
+
+    // IIRC we have 6 continuity pins
+    static int const PIN_COUNT = 6;
+    uint16_t pins[PIN_COUNT];
```
Comment (aidancostello) [inline MIDAS/src/sensors.h:32]: why do we have the pins array in the struct for the sensor as well as the struct for the sensor data? I don't think this array is necessary

File: MIDAS/src/systems.cpp [hunk 8]
```diff
@@ -85,6 +85,7 @@ DECLARE_THREAD(voltage, RocketSystems* arg) {
 DECLARE_THREAD(continuity, RocketSystems* arg) {
     while (true) {
         THREAD_SLEEP(16);
+        // How often do we check for continuity?
```
Comment (aidancostello) [inline MIDAS/src/systems.cpp:88]: 16 is fine for now, just add the continuity update function

**Review by aidancostello — CHANGES_REQUESTED** [review]

File: MIDAS/src/sensors.h (hunk 7 above)
Comment (redindelible) [reply MIDAS/src/sensors.h:32]: I think we like #defines more than static consts.

**Review by redindelible — CHANGES_REQUESTED** [review]
Same as Aidan

File: MIDAS/src/main.cpp (hunk 4 above)
Comment (zyunlam) [reply MIDAS/src/main.cpp:9]: This is my fault for trying to work on 3 tickets at the same time

File: MIDAS/src/hardware/Continuity.cpp [hunk 9]
```diff
@@ -1,17 +1,23 @@
...
-
-// global static instance of the sensor
+#include <hal.h>
 
+ADS7138_1Q ads7138_1q;
 
 ErrorCode ContinuitySensor::init() {
     // do whatever steps to initialize the sensor
     // if it errors, return the relevant error code
+    ads7138_1q.init();
```
Comment (aidancostello) [inline MIDAS/src/hardware/Continuity.cpp:11]: if this function returns true, return no error, if it returns false, we want to return a new error code that you'll have to make in error.h

**Review by aidancostello — CHANGES_REQUESTED** [review]
You also want to update the continuity thread function in systems.cpp so that you log the updated continuity struct.

**Review by aidancostello — APPROVED** [review]

**Review by redindelible — APPROVED** [review]

**Pull request merged**